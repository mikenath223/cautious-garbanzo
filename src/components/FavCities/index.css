.favCities {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-top: 2.5rem;
  list-style: none;
}
