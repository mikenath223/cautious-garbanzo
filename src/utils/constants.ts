import { IWeatherInfo } from "./types"

export const initFavCities: string[] = [
  "Tokyo",
  "Delhi",
  "Shanghai",
  "Dhaka",
  "Sao Paulo",
  "Mexico City",
  "Cairo",
  "Beijing",
  "Mumbai",
  "Osaka",
  "Chongqing",
  "Karachi",
  "Kinshasa",
  "Lagos",
  "Istanbul"
]

export const defaultCity: IWeatherInfo = {
  "request": {
    "type": "City",
    "query": "New York, United States of America",
  },
  "location": {
    "name": "New York",
    "country": "United States of America",
    "region": "New York",
    "localtime": "2019-09-07 08:14",
  },
  "current": {
    "observation_time": "12:14 PM",
    "temperature": 13,
    "weather_code": 113,
    "weather_icons": [
      "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
    ],
    "weather_descriptions": [
      "Sunny"
    ],
    "wind_speed": 0,
    "wind_degree": 349,
    "wind_dir": "N",
    "pressure": 1010,
    "precip": 0,
    "humidity": 90,
    "cloudcover": 0,
    "feelslike": 13,
    "uv_index": 4,
    "visibility": 16
  }
}

export const defaultNote = {
  id: "01",
  text: "Welcome"
}
