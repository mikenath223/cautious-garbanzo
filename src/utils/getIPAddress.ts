const fetchIPAddress = async () => {
  try {
    const response = await fetch('https://ipapi.co/json')
    const data = await response.json()
    return data
  } catch (err) {
    console.error(err)
  }
}

export default fetchIPAddress
