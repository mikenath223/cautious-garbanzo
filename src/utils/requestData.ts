import { IWeatherInfo, PromiseFulfilledResult } from "./types"

export const getCityWeather = async (city: string, isCheckCache: boolean): Promise<IWeatherInfo> => {
  try {
    const currentDate = new Date()
    const localStorageKey = `${city.toLowerCase().trim()}-${currentDate.getHours()}-${currentDate.getDate()}-${currentDate.getMonth()}-${currentDate.getFullYear()}`
    const cachedCityData = localStorage.getItem(localStorageKey)
    if (cachedCityData && isCheckCache) {
      const weatherData = JSON.parse(cachedCityData)
      return weatherData
    }
    const url = `http://api.weatherstack.com/current?access_key=04c73e2c81cd37fd215de627927cdd44&query=${city}`;
    const response = await fetch(url)
    const result = await response.json()
    if (result.error) {
      return result?.error?.info
    }
    localStorage.setItem(localStorageKey, JSON.stringify(result))
    return result
  } catch (error) {
    const errMsg = `Failed to retrieve ${city.trim()} data!`
    throw new Error(errMsg)
  }
}

const sortCities = (cities: string[]) => cities.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))

export const getFavsCityWeather = async (cities: string[]) => {
  const fetchRequests = sortCities(cities).map(city => getCityWeather(city, true));

  try {
    const citiesInfo = await Promise.allSettled(fetchRequests)
    const data = citiesInfo.filter(data => data.status === "fulfilled").map(data => (data as PromiseFulfilledResult<IWeatherInfo>).value)
    return data
  } catch (_) {
    return []
  }
}