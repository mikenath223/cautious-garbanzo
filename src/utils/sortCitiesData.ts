import { IWeatherInfo } from "./types";

const sortCitiesData = (cities: IWeatherInfo[] | undefined) => cities?.sort((a, b) => a.location.name.toLowerCase().localeCompare(b.location.name.toLowerCase()))

export default sortCitiesData
