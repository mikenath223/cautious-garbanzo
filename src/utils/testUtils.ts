import { render } from '@testing-library/react';

export const withRender = (component: JSX.Element) =>
  render(component);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => { };
  localStorage.getItem = (key: string) => null;
}