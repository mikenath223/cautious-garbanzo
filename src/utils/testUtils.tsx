import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export const withRender = (component: JSX.Element) =>
  render(<MemoryRouter>{component}</MemoryRouter>);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => {};
  localStorage.getItem = (key: string) => null;
};
