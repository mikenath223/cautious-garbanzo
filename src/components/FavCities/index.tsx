import { useCallback, useEffect } from "react";
import useLocalStorage from "../../hooks/useLocalStorage";
import { initFavCities } from "../../utils/constants";
import { getFavsCityWeather } from "../../utils/requestData";
import { IWeatherInfo } from "../../utils/types";
import FavCity from "../FavCity";
import sortCitiesData from "../../utils/sortCitiesData";
import "./index.css";

export default function FavCities() {
  const [favCitiesTemp, setFavCitiesTemp] = useLocalStorage<
    IWeatherInfo[] | undefined
  >("favorite-cities", undefined);

  const fetchCities = useCallback(async () => {
    let cities = await getFavsCityWeather(initFavCities);
    setFavCitiesTemp(cities);
  }, [setFavCitiesTemp]);

  useEffect(() => {
    if (!favCitiesTemp) {
      fetchCities();
    }
  }, [fetchCities, favCitiesTemp]);

  const onRemoveCity = (city: IWeatherInfo) => {
    let filteredCities = favCitiesTemp?.filter(
      (cityData) => cityData.location.name !== city.location.name
    );
    if (filteredCities) {
      let sortedCities = sortCitiesData(filteredCities);
      setFavCitiesTemp(sortedCities);
    }
  };

  return (
    <section>
      <ul className="favCities">
        {favCitiesTemp?.map((city: IWeatherInfo) => {
          let key = city.location.name;
          return <FavCity key={key} city={city} onRemoveCity={onRemoveCity} />;
        })}
      </ul>
      {favCitiesTemp?.length === 0 && (
        <h2>No favorite cities in list. Please add favorites.</h2>
      )}
    </section>
  );
}
