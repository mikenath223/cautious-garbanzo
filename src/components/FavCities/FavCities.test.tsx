import { withRender } from "../../utils/testUtils";
import FavCities from "./index";

describe("FavCities", function () {
  it("should match snapshot", () => {
    jest.spyOn(Storage.prototype, "setItem");
    Storage.prototype.setItem = jest.fn();
    jest.spyOn(Storage.prototype, "getItem");
    Storage.prototype.getItem = jest.fn();
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve({ test: 100 }),
      })
    ) as jest.Mock;
    const { container } = withRender(<FavCities />);
    expect(container).toMatchSnapshot();
  });
});
