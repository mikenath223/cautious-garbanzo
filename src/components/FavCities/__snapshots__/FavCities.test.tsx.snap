// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavCities should match snapshot 1`] = `
<div>
  <section>
    <ul
      class="favCities"
    />
  </section>
</div>
`;
